### PR TITLE
!dmrelay command

### PR DESCRIPTION
**Pull request by Den4200**
## Description
This PRs wipes out the old `DMRelay` cog and its functionalities. We realize that relaying all DMs sent to the bot into a specific channel may come in conflict with Discord's ToS, so that feature has now been removed, along with the `!reply` command. I've replaced it with the `!dmrelay` command, allowing moderators to view DM history between the bot and a specific user when necessary.

Here is an [example message log](https://paste.pythondiscord.com/ogeromerur.txt?noredirect).

For now, the message history is uploaded to our [paste site](https://paste.pydis.com). In the future, when [Logcord](https://github.com/python-discord/logcord) is up and ready, we will transfer over to use that.

### CODE REVIEW

File: bot/exts/moderation/dm_relay.py [hunk 1]
```diff
@@ -1,132 +1,63 @@
...
 
 log = logging.getLogger(__name__)
 
 
 class DMRelay(Cog):
-    """Relay direct messages to and from the bot."""
-
-    # RedisCache[str, t.Union[discord.User.id, discord.Member.id]]
-    dm_cache = RedisCache()
+    """Relay direct messages from the bot."""
```
Comment (jb3) [inline bot/exts/moderation/dm_relay.py:15]:
```suggestion
    """Inspect messages sent to the bot."""
```

File: bot/utils/services.py [hunk 2]
```diff
@@ -47,7 +47,13 @@ async def send_to_paste_service(contents: str, *, extension: str = "") -> Option
             continue
         elif "key" in response_json:
             log.info(f"Successfully uploaded contents to paste service behind key {response_json['key']}.")
-            return URLs.paste_service.format(key=response_json['key']) + extension
+
+            paste_link = URLs.paste_service.format(key=response_json['key']) + extension
+
+            if extension == '.py':
```
Comment (jb3) [inline bot/utils/services.py:53]: Shouldn't extension not have a `.` here?
```suggestion
            if extension == 'py':
```

File: bot/utils/services.py [hunk 3]
```diff
@@ -47,7 +47,13 @@ async def send_to_paste_service(contents: str, *, extension: str = "") -> Option
...
+                return paste_link
+            return paste_link + "?noredirect"
```
Comment (jb3) [inline bot/utils/services.py:55]:
```suggestion

            return paste_link + "?noredirect"
```

**Review by jb3 — CHANGES_REQUESTED** [review]
A few minor things noted in the code, as well as the below bugs.

## Does not handle client user

Screenshot 2021-03-28 at 12 40 57

We should gracefully reject requests to relay the bot's DM history with itself.

## Incorrect "no DM history"

After booting the bot and running this without sending any messages, the bot incorrectly identified me as having no history, after I sent a message to the (existing) channel it *did* identify that there was a history and uploaded it.

Maybe we should instead just use an async iterator to ensure we get a channel and if we yield 0 messages then we report back that there was no history, I presume this is a cache thing.

Screenshot 2021-03-28 at 12 41 30

File: bot/utils/services.py (hunk 2 above)
Comment (Den4200) [reply bot/utils/services.py:53]: A `.` is tacked onto the extension, if one is given beforehand.

https://github.com/python-discord/bot/blob/4be90b3c454138e3548c7394fcb2a1182b05b7d7/bot/utils/services.py#L22

Comment (jb3) [reply bot/utils/services.py:53]: Aha, gotcha.

**Review by jb3 — APPROVED** [review]
This looks good, however it now uses 2 API calls instead of one, which I think we can optimise down to one.

Switching around the order of "no channel" with "fetch messages" means you can do something along the lines of:
```py
found_messages = False

async for msg in user.history(...):
  found_messages = True

if not found_messages:
  await ctx.send("... no direct message hitsory ...")
```

I'm fine with this implementation though, since it's an infrequently used command anyway.

**Review by jb3 — APPROVED** [review]
I like it. :shipit:

File: bot/utils/services.py [hunk 4]
```diff
@@ -47,7 +47,14 @@ async def send_to_paste_service(contents: str, *, extension: str = "") -> Option
             continue
         elif "key" in response_json:
             log.info(f"Successfully uploaded contents to paste service behind key {response_json['key']}.")
-            return URLs.paste_service.format(key=response_json['key']) + extension
+
+            paste_link = URLs.paste_service.format(key=response_json['key']) + extension
+
+            if extension == '.py':
+                return paste_link
```
Comment (Numerlor) [inline bot/utils/services.py:54]: Does py need to be special cased here? the noredirect shouldn't change anything for it and the util isn't used anywhere where the compactness would matter

**Review by ChrisLovering — APPROVED** [review]
My only gripe is we're building the output string using `+=` rather than having a list and using `.join()` at the end. Not enough to stop the merge though! :D

File: bot/exts/moderation/dm_relay.py [hunk 5]
```diff
@@ -1,132 +1,68 @@
...
-                log.exception("Failed to send an attachment to the webhook")
-
-    async def cog_check(self, ctx: commands.Context) -> bool:
+        metadata = textwrap.dedent(f"""\
+            User: {user} ({user.id})
+            Channel ID: {user.dm_channel.id}\n
+        """)
+
+        paste_link = await send_to_paste_service(metadata + output, extension="txt")
+        await ctx.send(paste_link)
```
Comment (Numerlor) [inline bot/exts/moderation/dm_relay.py:59]: The case where the paste service fails and returns None needs to be handled here

**Review by ks129 — APPROVED** [review]

File: bot/exts/moderation/dm_relay.py [hunk 6]
```diff
@@ -1,132 +1,68 @@
...
-                    avatar_url=message.author.avatar_url
-                )
-            except discord.HTTPException:
-                log.exception("Failed to send an attachment to the webhook")
-
-    async def cog_check(self, ctx: commands.Context) -> bool:
+        metadata = textwrap.dedent(f"""\
+            User: {user} ({user.id})
+            Channel ID: {user.dm_channel.id}\n
+        """)
```
Comment (Numerlor) [inline bot/exts/moderation/dm_relay.py:56]:
```suggestion
        metadata = (
            f"User: {user} ({user.id})\n"
            f"Channel ID: {user.dm_channel.id}\n\n"
        )
```
Using implicit line continuation and joining looks a bit nicer to me; the textwrap import can also be removed if you decide to use this

File: bot/utils/services.py (hunk 4 above)
Comment (Den4200) [reply bot/utils/services.py:54]: I guess not, but I still think it's nice to account for it.

**Comment by Den4200** [comment]
I'll address these comments in a new PR.

**Pull request merged**